### DISCUSSION TITLE
Question: kernel patch from linux-tkg for ZFS

### DISCUSSION THREAD

**Discussion by mabod** [openzfs/zfs#11571] [Q&A] [answered]
This is a question for the ZFS developers.

While browsing for Linux custom kernels I stumbled over linux-tkg kernel. It contains a kernel patch which makes the symbols `kernel_fpu_begin` and `kernel_fpu_end` available for non-GPL modules again. 

https://github.com/Frogging-Family/linux-tkg/tree/master/linux-tkg-patches

Patch `0011-ZFS-fix.patch`:

```
From: =?UTF-8?q?J=C3=B6rg=20Thalheim?= <joerg@thalheim.io>
Date: Thu, 2 May 2019 05:28:08 +0100
Subject: [PATCH] x86/fpu: Export kernel_fpu_{begin,end}() with
 EXPORT_SYMBOL_GPL
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

We need these symbols in zfs as the fpu implementation breaks userspace:

https://github.com/zfsonlinux/zfs/issues/9346
Signed-off-by: Jörg Thalheim <joerg@thalheim.io>
---
 arch/x86/kernel/fpu/core.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/arch/x86/kernel/fpu/core.c b/arch/x86/kernel/fpu/core.c
index 12c70840980e..352538b3bb5d 100644
--- a/arch/x86/kernel/fpu/core.c
+++ b/arch/x86/kernel/fpu/core.c
@@ -102,7 +102,7 @@ void kernel_fpu_begin(void)
 	}
 	__cpu_invalidate_fpregs_state();
 }
-EXPORT_SYMBOL_GPL(kernel_fpu_begin);
+EXPORT_SYMBOL(kernel_fpu_begin);
 
 void kernel_fpu_end(void)
 {
@@ -111,7 +111,7 @@ void kernel_fpu_end(void)
 	this_cpu_write(in_kernel_fpu, false);
 	preempt_enable();
 }
-EXPORT_SYMBOL_GPL(kernel_fpu_end);
+EXPORT_SYMBOL(kernel_fpu_end);
 
 /*
  * Save the FPU state (mark it for reload if necessary):
-- 
2.23.0
```

Does it make sense to apply that patch from a ZFS standpoint of view? Does it have any performance benefits?

**Comment by behlendorf** [comment]
Since ZFS supports a wide range of Linux kernels, the compatibility code to detect and use these interfaces (when available) was kept.  Applying the patch may slightly improve performance since the kernel interfaces can be a bit more efficient, but the difference was quite small in the benchmarks we ran.

**Comment by AttilaFueloep** [comment, marked as answer]
Encryption performance wise it actually depends on the CPU and the encryption algorithm. If the CPU doesn't have AVX support or AES-CCM is chosen, that patch roughly doubles performance since FPU state is only saved on context switches. See e.g. #10846. If it has AVX and you are using AES-GCM Brian is right,  performance gain is quite minimal since recent 0.8 and 2.0 ZFS versions, besides other optimizations, mostly take care of that. See #9749.
